### PR TITLE
Openthread fix crypto psa import key

### PR DESCRIPTION
**Pull request by canisLupus1313**
According to PSA specification in case of `PSA_KEY_TYPE_ECC_KEY_PAIR`
function `psa_import_key` takes private key from key pair as argument.
This commit adds extraction of Private key from ECDSA key pair.

Also removes not needed `otPlatCryptoEcdsaGetPublicKey`.

### CODE REVIEW

File: modules/openthread/platform/crypto_psa.c [hunk 1]
```diff
@@ -134,13 +136,41 @@ otError otPlatCryptoImportKey(otCryptoKeyRef *aKeyRef,
...
-	psa_status_t status;
+	psa_status_t status = 0;
+	unsigned char *end;
 
 	if (aKeyRef == NULL || aKey == NULL || !checkKeyUsage(aKeyUsage)) {
 		return OT_ERROR_INVALID_ARGS;
 	}
 
+	/* Check if key is ECDSA pair and extract private key from it since PSA expects it. */
+	if (aKeyType == OT_CRYPTO_KEY_TYPE_ECDSA) {
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/crypto_psa.c:146]: should we guard these parts with `#if defined(CONFIG_OPENTHREAD_ECDSA)` ?

Comment (canisLupus1313) [reply modules/openthread/platform/crypto_psa.c:145]: fixed.

**Review by maciejbaczmanski — COMMENTED** [review]
few more suggestions / questions

File: modules/openthread/platform/crypto_psa.c [hunk 2]
```diff
@@ -121,26 +120,52 @@ void otPlatCryptoInit(void)
...
-			      otCryptoKeyStorage aKeyPersistence,
-			      const uint8_t *aKey,
+otError otPlatCryptoImportKey(otCryptoKeyRef *aKeyRef, otCryptoKeyType aKeyType,
+			      otCryptoKeyAlgorithm aKeyAlgorithm, int aKeyUsage,
+			      otCryptoKeyStorage aKeyPersistence, const uint8_t *aKey,
 			      size_t aKeyLen)
 {
+	int version;
+	size_t len;
+	unsigned char *p = (unsigned char *)aKey;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/crypto_psa.c:136]: I have wrapped those in `#if defined(CONFIG_OPENTHREAD_ECDSA)`  as for C convention declaration of variables sould be on top of function.

File: modules/openthread/platform/crypto_psa.c [hunk 3]
```diff
@@ -121,26 +120,52 @@ void otPlatCryptoInit(void)
...
-	psa_status_t status;
+	psa_status_t status = 0;
+	unsigned char *end;
 
 	if (aKeyRef == NULL || aKey == NULL || !checkKeyUsage(aKeyUsage)) {
 		return OT_ERROR_INVALID_ARGS;
 	}
-
+#if defined(CONFIG_OPENTHREAD_ECDSA)
+	/* Check if key is ECDSA pair and extract private key from it since PSA expects it. */
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/crypto_psa.c:148]: look at the comment above.

File: modules/openthread/platform/crypto_psa.c [hunk 4]
```diff
@@ -121,26 +120,52 @@ void otPlatCryptoInit(void)
...
+	if (aKeyType == OT_CRYPTO_KEY_TYPE_ECDSA) {
+
+		end = p + aKeyLen;
+		status = mbedtls_asn1_get_tag(&p, end, &len,
+					      MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE);
+		if (status != 0) {
+			return OT_ERROR_FAILED;
+		}
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/crypto_psa.c:156]: To keep with rest of the return types let's leave it as is.

File: modules/openthread/platform/crypto_psa.c [hunk 5]
```diff
@@ -121,26 +120,52 @@ void otPlatCryptoInit(void)
...
+
+		end = p + len;
+		status = mbedtls_asn1_get_int(&p, end, &version);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/crypto_psa.c:159]: No but it's needed in function call.

**Review by maciejbaczmanski — DISMISSED** [review]
lgtm

File: west.yml [hunk 6]
```diff
@@ -301,7 +301,7 @@ manifest:
       revision: da78aea63159771956fe0c9263f2e6985b66e9d5
       path: modules/lib/open-amp
     - name: openthread
-      revision: 4ed44bc7d58d9a98c6cca13a50d38129045ab3df
+      revision: pull/133/head
```
Comment (rlubos) [inline west.yml:304]: `00076aff3ae571db7c90509ec9dc293457098c35`

**Review by maciejbaczmanski — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by ArekBalysNordic — APPROVED** [review]

**Comment by Damian-Nordic** [comment]
@canisLupus1313 why did you remove `otPlatCryptoEcdsaGetPublicKey` function? Now the binary links the weak implementation from `crypto_platform.cpp` from OpenThread, based on mbedTLS. It's OK if it's not used right now but the platform API requires that it be provided.

**Comment by canisLupus1313** [comment]
@Damian-Nordic `otPlatCryptoEcdsaGetPublicKey` from plain text key pair is not used anymore in Openthread. The weak implementation was left there to indicate for users that this functionality shouldn't be used. That was the first reason.
Secondly the implementation of the function was never correct (since we haven't used it nobody noticed), the problem with is that even thought the  PSA function takes parameters that suggest it should be a key pair it really takes private key.

**Pull request merged**